Hyper/Game.cs

```diff
@@ -0,0 +1,168 @@
+            new VehiclesController(_scene, objectShader),
+            new LightSourcesController(_scene, lightSourceShader),
+            new HudController(windowHelper, hudShader),
+        };
+    }
+
+    public void Close()
+    {
+        foreach (var callback in _context.CloseCallbacks)
+        {
+            callback(_settings.SaveName);
+        }
+
+        _context.Clear();
+        foreach (var controller in _controllers)
+        {
+            controller.Dispose();
+        }
+        _scene.Dispose(); // Scene dispose needs to be after controller dispose.
+        _settings.Save();
+        LogManager.Flush();
+        IsRunning = false;
+    }
+
+    public void RenderFrame(FrameEventArgs e)
+    {
+        if (e.Time == 0) return;
+        if (!IsRunning) return;
+        GL.Clear(ClearBufferMask.ColorBufferBit | ClearBufferMask.DepthBufferBit);
+
+        foreach (var controller in _controllers)
+        {
+            controller.Render();
+        }
+    }
+
+    public void UpdateFrame(FrameEventArgs e)
+    {
+        if (e.Time == 0) return;
+        foreach (var callback in _context.FrameUpdateCallbacks)
+        {
+            callback(e);
+        }
+        _context.ExecuteAllHeldCallbacks(InputType.Key, e);
+        _context.ExecuteAllHeldCallbacks(InputType.MouseButton, e);
+    }
+
+    public void KeyDown(Keys key)
+    {
+        if (!_context.KeyDownCallbacks.ContainsKey(key))
+            return;
+
+        foreach (var callback in _context.KeyDownCallbacks[key])
+        {
+            callback();
+        }
+    }
+
+    public void KeyUp(Keys key)
+    {
+        if (!_context.KeyUpCallbacks.ContainsKey(key))
+            return;
+
+        foreach (var callback in _context.KeyUpCallbacks[key])
+        {
+            callback();
+        }
+    }
+
+    public void MouseMove(MouseMoveEventArgs e)
+    {
+        foreach (var callback in _context.MouseMoveCallbacks)
+        {
+            callback(e);
+        }
+    }
+
+    public void MouseDown(MouseButton button)
+    {
+        if (!_context.ButtonDownCallbacks.ContainsKey(button))
+            return;
+
+        foreach (var callback in _context.ButtonDownCallbacks[button])
+        {
+            callback();
+        }
+    }
+
+    public void MouseUp(MouseButton button)
+    {
+        if (!_context.ButtonUpCallbacks.ContainsKey(button))
+            return;
+
+        foreach (var callback in _context.ButtonUpCallbacks[button])
+        {
+            callback();
+        }
+    }
+
+    public void MouseWheel(MouseWheelEventArgs e)
+    {
+        _scene.Camera.Fov -= e.OffsetY;
+    }
+
+    public void Resize(ResizeEventArgs e)
+    {
+        GL.Viewport(0, 0, _size.X, _size.Y);
+        _scene.Camera.AspectRatio = _size.X / (float)_size.Y;
+        _size = e.Size;
+    }
+
+    public void RegisterCallbacks()
+    {
+        _context.RegisterKeys(new List<Keys> { Keys.Escape });
+
+        _context.RegisterKeyDownCallback(Keys.Escape, Close);
```

Comment:
hitting esc causes an exception to be thrown from ChunkWorker.RunJobs : `System.OperationCanceledException: 'The operation was canceled.'`

Reply:
I have not been able to reproduce this. Can you give some more details?

Reply:
I'm not getting this. are you using the latest version of this branch?

Reply:
To the best of my knowledge I am, steps to reproduce are literally the following:
1. Run Hyper project
2. Hit <kbd>Esc</kbd>
3.  `System.OperationCanceledException: 'The operation was canceled.'` (in ChunkWorker)

Reply:
should be fixed now

Reply:
confirmed